File: cpp/src/main/cpp/commands/ShootWithTableCommand.cpp
#include "commands/ShootWithTableCommand.h"

#include <cmath>

#include <frc/smartdashboard/SmartDashboard.h>

using namespace units::literals;

ShootWithTableCommand::ShootWithTableCommand(
    subsystems::ShooterSubsystem* shooter,
    subsystems::FeederSubsystem* feeder,
    std::function<shooting::ShotSetpoint()> setpointSupplier,
    std::function<bool()> alignmentReadySupplier, bool forcePitchHome)
    : shooter_(shooter),
      feeder_(feeder),
      setpoint_supplier_(std::move(setpointSupplier)),
      alignment_ready_supplier_(std::move(alignmentReadySupplier)),
      force_pitch_home_(forcePitchHome) {
  AddRequirements({shooter, feeder});
}

void ShootWithTableCommand::Initialize() {
  shot_timer_.Stop();
  shot_timer_.Reset();
  shot_timer_started_ = false;
  feeding_ = false;
  feeder_->Stop();

  if (force_pitch_home_ ||
      shooter_->GetPitchHomeState() ==
          subsystems::ShooterSubsystem::PitchHomeState::kUnhomed ||
      shooter_->GetPitchHomeState() ==
          subsystems::ShooterSubsystem::PitchHomeState::kFault) {
    shooter_->BeginPitchHoming();
  }
}

void ShootWithTableCommand::Execute() {
  if (!shooter_->IsPitchHomed()) {
    feeder_->SetBackwardFeederDuty(0.0);
    feeder_->SetUpwardDuty(0.0);
    return;
  }

  const shooting::ShotSetpoint setpoint = setpoint_supplier_();
  shooter_->ApplyShotSetpoint(setpoint);
  feeder_->SetUpwardFeederVelocity(setpoint.upperFeeder.value());

  if (!shot_timer_started_) {
    shot_timer_.Restart();
    shot_timer_started_ = true;
  }

  const bool ready =
      shooter_->IsFlywheelReady(0.7_tps) &&
      shooter_->IsPitchReady(0.75_deg) &&
      std::abs(feeder_->GetUpwardFeederVelocity() -
               setpoint.upperFeeder.value()) <= 2.0 &&
      alignment_ready_supplier_();

  if (shooting::ShotTable::FeedAllowed(
          shooter_->IsPitchHomed(), ready,
          shot_timer_.HasElapsed(1.5_s))) {
    feeding_ = true;
  }
  feeder_->SetBackwardFeederDuty(feeding_ ? 1.0 : 0.0);

  frc::SmartDashboard::PutBoolean("Shooting/Ready", ready);
  frc::SmartDashboard::PutBoolean("Shooting/ForcedFeed",
                                  feeding_ && !ready);
  frc::SmartDashboard::PutBoolean("Shooting/Feeding", feeding_);
}

void ShootWithTableCommand::End(bool interrupted) {
  shot_timer_.Stop();
  feeder_->Stop();
  shooter_->SetIdle();
}

bool ShootWithTableCommand::IsFinished() { return false; }

File: src/main/java/frc/robot/commands/ShootWithTableCommand.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.commands;

import java.util.function.BooleanSupplier;
import java.util.function.Supplier;

import edu.wpi.first.wpilibj.Timer;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.Command;

import frc.robot.shooting.ShotSetpoint;
import frc.robot.shooting.ShotTable;
import frc.robot.subsystems.FeederSubsystem;
import frc.robot.subsystems.ShooterSubsystem;

public class ShootWithTableCommand extends Command {
    private final ShooterSubsystem shooter;
    private final FeederSubsystem feeder;
    private final Supplier<ShotSetpoint> setpointSupplier;
    private final BooleanSupplier alignmentReadySupplier;
    private final boolean forcePitchHome;

    private boolean shotTimerStarted = false;
    private boolean feeding = false;
    private final Timer shotTimer = new Timer();

    public ShootWithTableCommand(
            ShooterSubsystem shooter,
            FeederSubsystem feeder,
            Supplier<ShotSetpoint> setpointSupplier,
            BooleanSupplier alignmentReadySupplier) {
        this(shooter, feeder, setpointSupplier, alignmentReadySupplier, false);
    }

    public ShootWithTableCommand(
            ShooterSubsystem shooter,
            FeederSubsystem feeder,
            Supplier<ShotSetpoint> setpointSupplier,
            BooleanSupplier alignmentReadySupplier,
            boolean forcePitchHome) {
        this.shooter = shooter;
        this.feeder = feeder;
        this.setpointSupplier = setpointSupplier;
        this.alignmentReadySupplier = alignmentReadySupplier;
        this.forcePitchHome = forcePitchHome;
        addRequirements(shooter, feeder);
    }

    @Override
    public void initialize() {
        shotTimer.stop();
        shotTimer.reset();
        shotTimerStarted = false;
        feeding = false;
        feeder.stop();

        if (forcePitchHome
                || shooter.getPitchHomeState() == ShooterSubsystem.PitchHomeState.UNHOMED
                || shooter.getPitchHomeState() == ShooterSubsystem.PitchHomeState.FAULT) {
            shooter.beginPitchHoming();
        }
    }

    @Override
    public void execute() {
        if (!shooter.isPitchHomed()) {
            feeder.setBackwardFeederDuty(0.0);
            feeder.setUpwardDuty(0.0);
            return;
        }

        ShotSetpoint setpoint = setpointSupplier.get();
        shooter.applyShotSetpoint(setpoint);
        feeder.setUpwardFeederVelocity(setpoint.feederRps());

        if (!shotTimerStarted) {
            shotTimer.restart();
            shotTimerStarted = true;
        }

        boolean ready = shooter.isFlywheelReady(0.7)
                && shooter.isPitchReady(0.75)
                && Math.abs(feeder.getUpwardFeederVelocity() - setpoint.feederRps()) <= 2.0
                && alignmentReadySupplier.getAsBoolean();

        if (ShotTable.feedAllowed(shooter.isPitchHomed(), ready, shotTimer.hasElapsed(1.5))) {
            feeding = true;
        }
        feeder.setBackwardFeederDuty(feeding ? 1.0 : 0.0);

        SmartDashboard.putBoolean("Shooting/Ready", ready);
        SmartDashboard.putBoolean("Shooting/ForcedFeed", feeding && !ready);
        SmartDashboard.putBoolean("Shooting/Feeding", feeding);
    }

    @Override
    public void end(boolean interrupted) {
        shotTimer.stop();
        feeder.stop();
        shooter.setIdle();
    }

    @Override
    public boolean isFinished() {
        return false;
    }
}
